src/posix/platform/rcp_caps_diag.cpp

```diff
@@ -498,22 +502,149 @@ void RcpCapsDiag::TestSpinelCommands(Category aCategory)
+{
+    static constexpr FlagEntry kRadioFlagEntries[] = {
+        {kCategoryThread1_1, OT_RADIO_CAPS_ACK_TIMEOUT},     {kCategoryThread1_1, OT_RADIO_CAPS_TRANSMIT_RETRIES},
+        {kCategoryThread1_1, OT_RADIO_CAPS_CSMA_BACKOFF},    {kCategoryThread1_2, OT_RADIO_CAPS_TRANSMIT_SEC},
+        {kCategoryThread1_2, OT_RADIO_CAPS_TRANSMIT_TIMING}, {kCategoryUtils, OT_RADIO_CAPS_ENERGY_SCAN},
+        {kCategoryUtils, OT_RADIO_CAPS_SLEEP_TO_TX},         {kCategoryUtils, OT_RADIO_CAPS_RECEIVE_TIMING},
+        {kCategoryUtils, OT_RADIO_CAPS_RX_ON_WHEN_IDLE}};
+
+    Output("\r\n%s :\r\n", CategoryToString(aCategory));
+
+    for (const FlagEntry &entry : kRadioFlagEntries)
+    {
+        if (entry.mCategory != aCategory)
+        {
+            continue;
+        }
+
+        OutputFormat(RadioCapbilityToString(entry.mFlag), SupportToString((aRadioCapFlags & entry.mFlag) > 0));
+    }
+}
+
+void RcpCapsDiag::TestSpinelCapbilityFlags(void)
+{
+    static constexpr uint8_t kCapsBufferSize = 100;
+    otError                  error;
+    uint8_t                  capsBuffer[kCapsBufferSize];
+    spinel_size_t            capsLength = sizeof(capsBuffer);
+
+    SuccessOrExit(error = mRadioSpinel.Get(SPINEL_PROP_CAPS, SPINEL_DATATYPE_DATA_S, capsBuffer, &capsLength));
+
+    Output("\r\nSpinel Capbility Flags :\r\n");
+    TestSpinelCapbilityFlags(kCategoryBasic, capsBuffer, capsLength);
+    TestSpinelCapbilityFlags(kCategoryUtils, capsBuffer, capsLength);
+
+exit:
+    if (error != OT_ERROR_NONE)
+    {
+        Output("Failed to get Spinel capbility flags: %s", otThreadErrorToString(error));
+    }
+
+    return;
+}
+
+void RcpCapsDiag::TestSpinelCapbilityFlags(Category aCategory, const uint8_t *aCapsData, spinel_size_t aCapsLength)
+{
+    static constexpr uint8_t   kCapsNameSize        = 40;
+    static constexpr FlagEntry kSpinelFlagEntries[] = {{kCategoryBasic, SPINEL_CAP_CONFIG_RADIO},
+                                                       {kCategoryBasic, SPINEL_CAP_MAC_RAW},
+                                                       {kCategoryBasic, SPINEL_CAP_RCP_API_VERSION},
+                                                       {kCategoryUtils, SPINEL_CAP_OPENTHREAD_LOG_METADATA},
+                                                       {kCategoryUtils, SPINEL_CAP_RCP_MIN_HOST_API_VERSION},
+                                                       {kCategoryUtils, SPINEL_CAP_RCP_RESET_TO_BOOTLOADER}};
+    char                       capName[kCapsNameSize];
+
+    Output("\r\n%s :\r\n", CategoryToString(aCategory));
+
+    for (const FlagEntry &entry : kSpinelFlagEntries)
+    {
+        if (entry.mCategory != aCategory)
+        {
+            continue;
+        }
+
+        snprintf(capName, sizeof(capName), "SPINEL_CAPS_%s", spinel_capability_to_cstr(entry.mFlag));
+        OutputFormat(capName, SupportToString(IsSpinelCapabilitySupported(aCapsData, aCapsLength, entry.mFlag)));
+    }
+}
```

Comment:
What about doing like this:
```
void RcpCapsDiag::TestSpinelCapbilityFlags(FlagEntry *aFlagEntries, uint8_t aEntriesLength, const uint8_t *aCapsData, spinel_size_t aCapsLength)
{
     // Simply iterate over all flag entries and do the output.
}
```
And in `TestSpinelCapbilityFlags`:
```
void RcpCapsDiag::TestSpinelCapbilityFlags(void)
{
    static constexpr FlagEntry kSpinelFlagEntriesBasic[] = {{kCategoryBasic, SPINEL_CAP_CONFIG_RADIO},
                                                           {kCategoryBasic, SPINEL_CAP_MAC_RAW},
                                                           {kCategoryBasic, SPINEL_CAP_RCP_API_VERSION}};
    static constexpr FlagEntry kSpinelFlagEntriesUtils[] = {kCategoryUtils, SPINEL_CAP_OPENTHREAD_LOG_METADATA},
                                                           {kCategoryUtils, SPINEL_CAP_RCP_MIN_HOST_API_VERSION},
                                                           {kCategoryUtils, SPINEL_CAP_RCP_RESET_TO_BOOTLOADER}};
    ...
    TestSpinelCapbilityFlags(kSpinelFlagEntriesBasic, sizeof(kSpinelFlagEntriesBasic), ...);
    TestSpinelCapbilityFlags(kSpinelFlagEntriesUtils, sizeof(kSpinelFlagEntriesUtils), ...);

    return;
}
```
In this way, we can skip the category comparison. And we can even remove the category field in `FlagEntry`.

Just a suggestion and it's up to you.

Reply:
Updated